proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+This allows us to rationalize our `Span<T>` definition as well as allowing third party customers rationalize their existing APIs. 
+
+```c#
+// Existing Span<T> definition
+public readonly ref struct Span<T>
+{
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array, int index, int length) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public static implicit operator Span<T>(T[]? array) { ... }
+}
+```
+
+This does not impact the binding rules for invocations of such members. These symbols will continue to be fully bindable. In the case they are bound when any of the type parameters on the containing type are potentially a `ref struct` then use of a member annotated with `[SupportsOnlyNonByRefLike]` will be considered an error.
+
+```c#
+void M<T, U>()
+    where T : ref struct
+{
+    // Error: cannot access member as `T` is potentially a `ref struct`
+    new Span<T>(null);
+
+    // Okay
+    new Span<U>(null);
+}
+```
+
+The runtime will recognize this attribute and ensure that any use of these members at runtime with a type that is a `ref struct` will result in a runtime exception.
+
+Note: the [alternative](#special-case) here is simply special casing `Span<T>` as array is special cased today.
+
+### ref struct interfaces
+<a name="interfaces"></a>
+
+The language will allow `ref struct` to implement interfaces. The rules for implementation will be the same as for normal `struct`. This includes allowing the implementations to be implicit or explicit.
+
+**TODO: Need better example than animals but feeling lazy**
+
+```c#
+interface IAnimal
+{
+    void Speak(); 
+}
+
+ref struct Lab : IAnimal
+{
+    public void Speak() => Console.WriteLine("FEED ME");
+}
+```
+
+The one exception is when `ref struct` intersects with default implementation methods. When interfaces containing default implemented methods are used on `struct` types there is an implicit boxing operation for the invocation. That means effectively a DIM call when the implementing type was a `ref struct` would violate our rules on them appearing in the heap. As such this will be disallowed by both the language and runtime. 
+
+- A `ref struct` that implements an `interface` must provide implementations for all methods, even those with a default implementation
+- A DIM invocation on a `ref struct` will result in a runtime exception
+
+The ability to implement interfaces does not change the rules around conversions though. A `ref struct` cannot be boxed, conversion from a `struct` to an `interface` has an implicit box associated with it hence conversion from `ref struct` to its implemented interfaces remains an error: 
```

Comment:
```suggestion
The ability to implement interfaces does not change the rules around conversions though. A `ref struct` cannot be boxed, and since conversion from a `struct` to an `interface` has an implicit box associated with it, conversion from a `ref struct` to its implemented interfaces remains an error: 
```